keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,639 @@
+1. The Sidecar detects the new `BucketContent` object and gets the associated `BucketClass`.
+1. The Sidecar calls the CreateBucket() rpc, passing the `bucketClass.parameters` and is returned a bucket endpoint and credentials.
+1. The Sidecar creates a secret containing the endpoint and credentials, with a random/unique name and `ownerRef` set to `BucketContent`.
+1. The Sidecar updates `BucketContent.secretRef` with its `secret` name and namespace and sets `BucketContent.status.phase` to *“Ready”.*
+1. The Controller detects the `BucketContent` update and sees the *“Ready”* phase. 
+1. The Controller copies the generated `secret` to the `Bucket` namespace with name defined in `Bucket.secretName`.  The `secret` is created with an `ownerRef` of the `Bucket`.
+1. The Controller “binds” the `Bucket` and `BucketContent` by setting `bucket.status.bucketContentName` and `bucketContent.bucketRef`, and sets both statuses to *“Bound”*.
+1. The app `pod` ingests `secret` and runs.
+
+#### Grant Bucket Access (Brownfield)
+
+1. User creates a `Bucket` in their namespace, with reference to a `BucketClass`.
+1. Controller sees the new `Bucket` and applies a `finalizer` for orchestrated deletions.
+1. Controller gets the `BucketClass` referenced by the `Bucket`.
+1. Controller creates a `BucketContent` object with its `BucketClassName` set to the name of its `BucketClass` and a `finalizer`.
+1. Sidecar detects the new `BucketContent` object and gets the associated `BucketClass`.
+1. Sidecar calls the GrantBucketAccess() rpc, passing the `bucketClass.bucketIdentifier` and the `bucketClass.parameters` and is returned a bucket endpoint and credentials.
+1. Sidecar creates a `secret` containing the endpoint and credentials, with a random/unique name and `ownerRef` set to `BucketContent`.
+1. Sidecar updates `BucketContent.secretRef` with its `secret` name and namespace and sets `BucketContent.status.phase` to *“Ready”*.
+1. Controller detects the `BucketContent` update and sees the *“Ready”* phase. 
+1. Controller copies the generated `secret` to the `Bucket` namespace with name `Bucket.secretName`.  The `secret` is created with an ownerRef of the Bucket.
+1. Controller *“binds”* the `Bucket` and `BucketContent` by setting `bucket.status.bucketContentName` and `bucketContent.bucketRef`, and sets both statuses to *“Bound”*.
+1. App `pod` ingests `secret` and runs.
+
+#### Delete Or Revoke Access (Greenfield & Brownfield)
+
+1. The user deletes their `Bucket`, which blocks until the `finalizer` is removed.
+1. The Controller detects the event and sees the `deletionTimestamp` set in `Bucket` and gets the `BucketClass`. 
+1. If the `BucketClass.secretRef` is nil, the object store bucket is not static, and the process continues to step 4.
+1. The Controller deletes the referenced `BucketContent` object, which blocks until the `finalizer` is removed.
```

Comment:
Deleting a bucket can be a long-running operation if there are many objects in the bucket. We should mark the underlying bucket for deletion and allow it to be removed asynchronously IMO. Is this the proposal, or would it be up to each driver to implement whether it blocks until the bucket and all its contents are removed?

Reply:
I think it would be up to each driver (assuming we implement earlier versions of this KEP where we discussed automation implementation ideas). But this is an interesting point and probably needs further discussion after MVP. @travisn do you think API changes will be needed if we want to support async delete later on?